typhos/positioner.py

```diff
@@ -122,90 +228,183 @@ def _get_position(self):
             raise Exception("No Device configured for widget!")
         return self._readback.get()
 
+    @_linked_attribute('readback_attribute', 'ui.user_readback')
+    def _link_readback(self, signal, widget):
+        """Link the positioner readback with the ui element."""
+        self._readback = signal
+
+    @_linked_attribute('setpoint_attribute', 'ui.user_setpoint')
+    def _link_setpoint(self, signal, widget):
+        """Link the positioner setpoint with the ui element."""
+        self._setpoint = signal
+        if signal is not None:
+            # Seed the set_value text with the user_setpoint channel value.
+            if hasattr(widget, 'textChanged'):
+                widget.textChanged.connect(self._user_setpoint_update)
+
+    @_linked_attribute('low_limit_switch_attribute', 'ui.low_limit_switch')
+    def _link_low_limit_switch(self, signal, widget):
+        """Link the positioner lower limit switch with the ui element."""
+        if signal is None:
+            widget.hide()
+
+    @_linked_attribute('high_limit_switch_attribute', 'ui.high_limit_switch')
+    def _link_high_limit_switch(self, signal, widget):
+        """Link the positioner high limit switch with the ui element."""
+        if signal is None:
+            widget.hide()
+
+    @_linked_attribute('low_limit_switch_attribute', 'ui.low_limit')
+    def _link_low_limit(self, signal, widget):
+        """Link the positioner lower limit with the ui element."""
+        return signal is not None
+
+    @_linked_attribute('high_limit_switch_attribute', 'ui.high_limit')
```

Comment:
I think there is a typo in here and above, since the soft limit positions aren't making it into the UI (instead, the 1 or 0 for hit/not hit are used). But after swapping this in there are exceptions raised.
```suggestion
    @_linked_attribute('high_limit_attribute', 'ui.high_limit')
```

Reply:
Great catch. My eyes were going buggy reading the similar string/method names. Pushed a fix.

Can you give me details on the exception? Mine is working with EpicsMotor, at least.

Reply:
Using `typhos "pcdsdevices.device_types.BeckhoffAxis[{'prefix': 'PLC:TST:MOT:SIM:03'}]"` on psbuil-rhel7-01 and switching to the positioner template:
```
[2020-05-14 13:32:30] - ERROR - device.low_limit => self.ui.low_limit (signal: 0.0 widget: <pydm.widgets.label.PyDMLabel object at 0x7f84380ac1f8>)
Traceback (most recent call last):
  File "/reg/neh/home/zlentz/pydev/typhos/positioner.py", line 70, in wrapped
    _link_signal_to_widget(signal, widget)
  File "/reg/neh/home/zlentz/pydev/typhos/positioner.py", line 27, in _link_signal_to_widget
    plugins.register_signal(signal)
  File "/reg/neh/home/zlentz/pydev/typhos/plugins/core.py", line 27, in register_signal
    if signal.name in signal_registry:
AttributeError: 'float' object has no attribute 'name'
```

Reply:
Ah, OK, what's happening there is that it's trying to link `device.low_limit` to the UI element `.low_limit`, but `device.low_limit = float()` - likely a property?

I envision this being fixed by per-class templates which indicate the correct attributes. Depending on just how annoying that ends up being, we might circle back and be more consistent in our component names? Or `[alternate solution here]`...

Reply:
https://github.com/pcdshub/pcdsdevices/blob/f153abeea529d4eaa29e5cd786a24a39dd0dc210/pcdsdevices/epics_motor.py#L65-L70

Reply:
Think this is a reasonable path?

Reply:
Got it, so for `EpicsMotor` subclasses we'd set that property to `low_travel_limit`, for example?

Reply:
https://github.com/bluesky/ophyd/blob/eee8797e79448a1dc02c65952af475563f4d260a/ophyd/epics_motor.py#L67

Reply:
Actually, it seems to me that `low_limit` is always a property, even in `PositionerBase`: https://github.com/bluesky/ophyd/blob/eee8797e79448a1dc02c65952af475563f4d260a/ophyd/positioner.py#L143

Reply:
Ohh, good! That actually helps quite a bit. I'll change the default to that `low_limit_travel` then.

So it'll use that, if available, or fall back to a static `low_limit, high_limit = obj.limits` query.

Reply:
Resolved in bad30dc